File: modules/ui/src/main/java/eu/europa/ec/leos/ui/extension/SliderPinsExtension.java
/*
 * Copyright 2019 European Commission
 *
 * Licensed under the EUPL, Version 1.2 or – as soon they will be approved by the European Commission - subsequent versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 *     https://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */
package eu.europa.ec.leos.ui.extension;

import com.vaadin.annotations.JavaScript;
import com.vaadin.annotations.StyleSheet;
import com.vaadin.data.HasValue;
import com.vaadin.shared.Registration;
import com.vaadin.ui.*;
import eu.europa.ec.leos.web.support.LeosCacheToken;

import java.util.Map;
import java.util.WeakHashMap;

@StyleSheet({"vaadin://../js/lib/sliderPins/css/sliderPins.css" + LeosCacheToken.TOKEN,
             "vaadin://../assets/css/leos-sliderPins.css"+ LeosCacheToken.TOKEN})
@JavaScript({"vaadin://../js/ui/extension/SliderPinsConnector.js"+ LeosCacheToken.TOKEN })
public class SliderPinsExtension<T extends AbstractComponent> extends LeosJavaScriptExtension {

    private static final long serialVersionUID = 1L;
    
    private Map<Component, Registration> listenersMap;

    public SliderPinsExtension(T target, Map<String, String> configMap){
        super();
        listenersMap = new WeakHashMap();
        getState().configMap = configMap;
        extend(target);
    }

    protected void extend(T target) {
        super.extend(target);
        // handle target's value change
        if (target instanceof ComponentContainer){
            ComponentContainer layout = (ComponentContainer) target;
            //FIXME If the field value is set before adding the component the "ValueChangeListener" has no effect
            layout.addComponentAttachListener(event -> {
                addValueChangeListener(event.getAttachedComponent());
            });
            layout.addComponentDetachListener(event -> {
                Component detachedComponent = event.getDetachedComponent();
                if (listenersMap.containsKey(detachedComponent)) {
                    listenersMap.get(detachedComponent).remove();
                    listenersMap.remove(detachedComponent);
                }
            });
            
            layout.forEach(component -> addValueChangeListener(component));
        }
        else {
            addValueChangeListener(target);
        }
    }

    private void addValueChangeListener(Component component) {
        if (component instanceof HasValue) {
            Registration valueChangeListener = ((HasValue) component).addValueChangeListener(event -> {
                LOG.trace("Target's value changed...");
                // Mark that this connector's state might have changed.
                // There is no need to send new data to the client-side,
                // since we just want to trigger a state change event...
                forceDirty();
            });
            listenersMap.put(component, valueChangeListener);
        }
    }

    public void navigateSliderPins(String direction) {
        callFunction("navigateSliderPins", direction);
    }
    
    @Override
    protected SliderPinsState getState() {
        return (SliderPinsState) super.getState();
    }

    @Override
    protected SliderPinsState getState(boolean markAsDirty) {
        return (SliderPinsState) super.getState(markAsDirty);
    }
}

File: modules/js/src/main/js/ui/extension/SliderPinsExtension.js
/*
 * Copyright 2019 European Commission
 *
 * Licensed under the EUPL, Version 1.2 or – as soon they will be approved by the European Commission - subsequent versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 *     https://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */
; // jshint ignore:line
define(function SliderPinsExtensionModule(require) {
    "use strict";

    // load module dependencies
    var log = require("logger");
    var sliderPins = require("sliderPins");
    var UTILS = require("core/leosUtils");

    function _initPins(connector) {
        log.debug("Initializing Slider Pins extension...");

        // restrict scope to the extended target
        connector.target = UTILS.getParentElement(connector);

        _updateParentPosition(connector.target);
        sliderPins.create(connector.target, connector.getState().configMap);

        log.debug("Registering Slider Pins unregistration listener...");
        connector.onUnregister = _connectorUnregistrationListener;

        log.debug("Registering Slider Pins state change listener...");
        connector.onStateChange = _connectorStateChangeListener;
        
        connector.navigateSliderPins = _navigate;
    }


    function _updateParentPosition(target){
        target.parentNode.style.position="relative";
    }

    // handle connector unregistration on client-side
    function _connectorStateChangeListener() {
        var connector = this;
        log.debug("Slider Pins extension state changed...");
        sliderPins.destroy(connector.target);
        // KLUGE delay execution due to sync issues with target update
        setTimeout(sliderPins.create, 500, connector.target, connector.getState().configMap);
    }

    // handle connector unregistration on client-side
    function _connectorUnregistrationListener() {
        var connector = this;
        log.debug("Unregistering Slider Pins extension...");
        sliderPins.destroy(connector.target);
        connector.target = null;
    }

    function _navigate(direction) {
        var connector = this;
        sliderPins.navigate(connector.target, direction);
    }
    
    return {
        init: _initPins
    };
});
